### ISSUE TITLE
salt winrepo - jinja rendering adds spurious extra trailing space

### ISSUE THREAD

**Issue by TheBigBear** [saltstack/salt#26159]
I have across a winrepo jinja rendering bug. Under some situations it adds a 'trailing' space at end of rendered string.

I had problems with uninstaller entries of the form
```
{% set Program_Files = "C:\Program Files" %}
...
uninstaller: '{{ Program_Files }}\myprog\myuninst.exe'
```
Giving me 'broken' cmd lines that had trailing spaces after the jinja rendered part and then I realised that the whole string wasn't quoted. So I tried improving it to:
```
{% set Program_Files = "C:\Program Files (x86)" %}
...
uninstaller: '"{{ Program_Files }}\myprog\myuninst.exe"'
```

now resulting in a error 5 - access denied error message?
```
 Error running 'pkg.remove': Unable to run command ['"C:\\Program Files \\CCleaner\\uninst.exe"', '/S'] with the context {'with_communicate': True, 'shell': False, 'env': {'TMP': 'C:\\Users\\sysadmin.uk\\AppData\\Local\\Temp', 'COMPUTERNAME': 'UK-IT-DEV1', 'USERDOMAIN': 'GLOBAL', 'PYTHON': 'C:\\salt\\bin\\python.exe', 'PSMODULEPATH': 'C:\\Users\\sysadmin.uk\\Documents\\WindowsPowerShell\\Modules;C:\\Users\\sysadmin.uk\\AppData\\Roaming\\Boxstarter', 'COMMONPROGRAMFILES': 'C:\\Program Files\\Common Files', 'PROCESSOR_IDENTIFIER': 'Intel64 Family 6 Model 70 Stepping 1, GenuineIntel', 'PROGRAMFILES': 'C:\\Program Files', 'PROCESSOR_REVISION': '4601', 'SYSTEMROOT': 'C:\\Windows', 'PATH': 'C:\\Program Files (x86)\\GnuWin32\\bin;C:\\Program Files (x86)\\Notepad++;C:\\ProgramData\\chocolatey\\bin;c:\\sysutl;c:\\utl;C:\\Windows;C:\\Windows\\system32;C:\\Windows\\System32\\Wbem;C:\\Windows\\System32\\WindowsPowerShell\\v1.0;c:\\python27;c:\\python27\\scripts;C:\\Program Files (x86)\\Skype\\Phone\\;C:\\Users\\sysadmin.uk\\AppData\\Roaming\\Boxstarter;C:\\salt\\bin\\lib\\site-packages\\pywin32_system32', 'PROGRAMFILES(X86)': 'C:\\Program Files (x86)', 'TEMP': 'C:\\Users\\sysadmin.uk\\AppData\\Local\\Temp', 'SCRIPT': 'C:\\salt\\bin\\Scripts\\salt-call', 'COMMONPROGRAMFILES(X86)': 'C:\\Program Files (x86)\\Common Files', 'PROCESSOR_ARCHITECTURE': 'AMD64', 'CHOCOLATEYINSTALL': 'C:\\ProgramData\\chocolatey', 'ALLUSERSPROFILE': 'C:\\ProgramData', 'SALTDIR': 'C:\\salt', 'LOCALAPPDATA': 'C:\\Users\\sysadmin.uk\\AppData\\Local', 'HOMEPATH': '\\Users\\sysadmin.uk', 'USERDOMAIN_ROAMINGPROFILE': 'GLOBAL', 'PROGRAMW6432': 'C:\\Program Files', 'USERNAME': 'sysadmin.uk', 'LOGONSERVER': '\\\\UK-DC1', 'PROMPT': '$P$G', 'COMSPEC': 'C:\\Windows\\system32\\cmd.exe', 'PROGRAMDATA': 'C:\\ProgramData', 'USERDNSDOMAIN': 'GLOBAL.LOCAL', 'PATHEXT': '.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC', 'FP_NO_HOST_CHECK': 'NO', 'WINDIR': 'C:\\Windows', 'OPENSSL_CONF': 'C:\\OpenSSL-Win64\\bin\\openssl.cfg', 'HOMEDRIVE': 'C:', 'APPDATA': 'C:\\Users\\sysadmin.uk\\AppData\\Roaming', 'SYSTEMDRIVE': 'C:', 'NUMBER_OF_PROCESSORS': '1', 'PROCESSOR_LEVEL': '6', 'COMMONPROGRAMW6432': 'C:\\Program Files\\Common Files', 'OS': 'Windows_NT', 'PUBLIC': 'C:\\Users\\Public', 'USERPROFILE': 'C:\\Users\\sysadmin.uk'}, 'stdout': -1, 'stdin': None, 'stderr': -2, 'cwd': 'C:\\Users\\sysadmin.uk'}, reason: [Error 5] Access is denied
```
Ahh, just had another look. It still has a trailing 'space' right after the ```"Program Files"```
```C:\\Program Files \\CCleaner...``` part
```          extra trail space                 ^```

to reproduce try with this version of ccleaner.sls https://github.com/saltstack/salt-winrepo/blob/5b4805e7c0ed449a188d4b0390e1ee5497a66565/ccleaner.sls

@twangboy @jfindlay @UtahDave any ideas on where to fix this?

**Comment by TheBigBear** [comment]
To illustrate this issue further and add another strange angle on it. ;-)
If I do the pkg.remove on a ```AMD64``` ```cpuarch``` machine it adds the trailing extra 'space', if I do the same action from same sls file on ```x86``` ```cpuarch``` machine it does not?

__So the jinja rendering of the same sls file is different on the two platforms?__


```salt-call --local pkg.remove ccleaner out=yaml`` when run on a x86 cpuarch machine.
produces:
```
[WARNING ] C:\salt\bin\lib\site-packages\salt\modules\cmdmod.py:241: RuntimeWarning: tempnam is a potential security risk to your program
  cwd = os.tempnam()[:3]

[INFO    ] Executing command ['"C:\\Program Files\\CCleaner\\uninst.exe"', '/S'] in directory 'C:\\'
Error running 'pkg.remove': Unable to run command ['"C:\\Program Files\\CCleaner\\uninst.exe"', '/S'] with the context {'with_communicate': True, 'shell': False, 'env': {'TMP': 'C:\\Users\\sysadmin\\AppData\\Local\\Temp', 'COMPUTERNAME': 'UK-IT-70', 'USERDOMAIN': 'WORKGROUP', 'VS100COMNTOOLS': 'C:\\Program Files\\Microsoft Visual Studio 10.0\\Common7\\Tools\\', 'P7MTOOLDIR': 'C:\\Program Files\\Cryptigo\\p7mTool', 'PSMODULEPATH': 'C:\\Windows\\system32\\WindowsPowerShell\\v1.0\\Modules\\', 'COMMONPROGRAMFILES': 'C:\\Program Files\\Common Files', 'PROCESSOR_IDENTIFIER': 'x86 Family 6 Model 70 Stepping 1, GenuineIntel', 'PROGRAMFILES': 'C:\\Program Files', 'PROCESSOR_REVISION': '4601', 'SYSTEMROOT': 'C:\\Windows', 'HOME': 'C:\\Users\\ursr', 'SVN_SSH': 'C:\\Program Files\\TortoiseGit\\bin\\TortoisePlink.exe', 'GIT_SSH': 'C:\\Program Files\\TortoiseGit\\bin\\TortoisePlink.exe', 'TEMP': 'C:\\Users\\sysadmin\\AppData\\Local\\Temp', 'SCRIPT': 'C:\\salt\\bin\\Scripts\\salt-call', 'PROCESSOR_ARCHITECTURE': 'x86', 'NEWLISPDIR': 'C:\\Program Files\\newlisp', 'PYTHON': 'C:\\salt\\bin\\python.exe', 'SALTDIR': 'C:\\salt', 'LOCALAPPDATA': 'C:\\Users\\sysadmin\\AppData\\Local', 'HOMEPATH': '\\Users\\sysadmin', 'USERNAME': 'sysadmin.uk', 'LOGONSERVER': '\\\\LINUX', 'PROMPT': '$P$G', 'COMSPEC': 'C:\\Windows\\system32\\cmd.exe', 'PROGRAMDATA': 'C:\\ProgramData', 'ALLUSERSPROFILE': 'C:\\ProgramData', 'PATH': 'c:\\OpenJade-1.3\\bin;c:\\gtk\\bin;C:\\Program Files\\Common Files\\Microsoft Shared\\Windows Live;C:\\Program Files\\Notepad++;C:\\Ruby193\\bin;C:\\Program Files\\Windows Resource Kits\\Tools\\;C:\\Program Files\\ZBar\\bin;C:\\Program Files\\ActiveState Komodo Edit 6\\;C:\\Program Files\\VMware\\VMware vSphere CLI\\Perl\\bin;C:\\Program Files\\VMware\\VMware vSphere CLI\\Perl\\site\\bin;C:\\Program Files\\VMware\\VMware VI Remote CLI\\Perl\\bin;C:\\Windows\\system32;C:\\Windows;C:\\Windows\\System32\\Wbem;C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\;c:\\sysutl;c:\\utl;C:\\vim\\vim73;C:\\Program Files\\WinMerge;C:\\Program Files\\GNU\\GnuPG\\pub;C:\\Program Files\\Notepad++;C:\\Windows\\System32\\Windows System Resource Manager\\bin;C:\\Windows\\idmu\\common;C:\\ErgoEmacs\\bin;C:\\ErgoEmacs;C:\\Program Files\\Windows Live\\Shared;C:\\Program Files\\Perforce\\;C:\\Program Files\\Network Monitor Experts\\TCP Analyzer 1.2\\;C:\\Program Files\\Microsoft Windows Performance Toolkit\\;C:\\Program Files\\VMware\\VMware VI Remote CLI\\Perl\\site\\bin;C:\\Program Files\\Git\\cmd;C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\;C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\;C:\\Program Files\\PuTTY;C:\\Program Files\\OpenVPN\\bin;C:\\Program Files\\Nmap;C:\\Program Files\\proXPN\\bin;C:\\salt\\bin\\lib\\site-packages\\pywin32_system32', 'PATHEXT': '.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC;.py;.pyw;.RB;.RBW', 'ASL.LOG': 'Destination=file;OnFirstLog=command,environment,parent', 'UD_INSTALL_DIR': 'C:\\Program Files\\UltraDefrag', 'FP_NO_HOST_CHECK': 'NO', 'WINDIR': 'C:\\Windows', 'SQL_CLIENT_CHARSET': '1252', 'HOMEDRIVE': 'C:', 'SYSTEMDRIVE': 'C:', 'NUMBER_OF_PROCESSORS': '2', 'APPDATA': 'C:\\Users\\sysadmin\\AppData\\Roaming', 'PROCESSOR_LEVEL': '6', 'OS': 'Windows_NT', 'PUBLIC': 'C:\\Users\\Public', 'USERPROFILE': 'C:\\Users\\sysadmin'}, 'stdout': -1, 'stdin': None, 'stderr': -2, 'cwd': 'C:\\'}, reason: [Error 5] Access is denied

Note far right end of line above reads:

reason: [Error 5] Access is denied
```

But if I run the same cmd on a ```AMD64``` ```cpuarch``` machine, I get it __with that spurious extra trailing 'space'__ after the "{{ Program_Files }}" jinja rendered part. Why?

```
[INFO    ] Executing command ['"C:\\Program Files \\CCleaner\\uninst.exe"', '/S'] in directory 'C:\\Users\\sysadmin.uk'
Error running 'pkg.remove': Unable to run command ['"C:\\Program Files \\CCleaner\\uninst.exe"', '/S'] with the context {'with_communicate': True, 'shell': False, 'env': {'TMP': 'C:\\Users\\sysadmin.uk\\AppData\\Local\\Temp', 'COMPUTERNAME': 'UK-IT-DEV1', 'USERDOMAIN': 'GLOBAL', 'PYTHON': 'C:\\salt\\bin\\python.exe', 'PSMODULEPATH': 'C:\\Users\\sysadmin.uk\\Documents\\WindowsPowerShell\\Modules;C:\\Users\\sysadmin.uk\\AppData\\Roaming\\Boxstarter', 'COMMONPROGRAMFILES': 'C:\\Program Files\\Common Files', 'PROCESSOR_IDENTIFIER': 'Intel64 Family 6 Model 70 Stepping 1, GenuineIntel', 'PROGRAMFILES': 'C:\\Program Files', 'PROCESSOR_REVISION': '4601', 'SYSTEMROOT': 'C:\\Windows', 'PATH': 'C:\\Program Files (x86)\\GnuWin32\\bin;C:\\Program Files (x86)\\Notepad++;C:\\ProgramData\\chocolatey\\bin;c:\\sysutl;c:\\utl;C:\\Windows;C:\\Windows\\system32;C:\\Windows\\System32\\Wbem;C:\\Windows\\System32\\WindowsPowerShell\\v1.0;c:\\python27;c:\\python27\\scripts;C:\\Program Files (x86)\\Skype\\Phone\\;C:\\Users\\sysadmin.uk\\AppData\\Roaming\\Boxstarter;C:\\salt\\bin\\lib\\site-packages\\pywin32_system32', 'PROGRAMFILES(X86)': 'C:\\Program Files (x86)', 'TEMP': 'C:\\Users\\sysadmin.uk\\AppData\\Local\\Temp', 'SCRIPT': 'C:\\salt\\bin\\Scripts\\salt-call', 'COMMONPROGRAMFILES(X86)': 'C:\\Program Files (x86)\\Common Files', 'PROCESSOR_ARCHITECTURE': 'AMD64', 'CHOCOLATEYINSTALL': 'C:\\ProgramData\\chocolatey', 'ALLUSERSPROFILE': 'C:\\ProgramData', 'SALTDIR': 'C:\\salt', 'LOCALAPPDATA': 'C:\\Users\\sysadmin.uk\\AppData\\Local', 'HOMEPATH': '\\Users\\sysadmin.uk', 'USERDOMAIN_ROAMINGPROFILE': 'GLOBAL', 'PROGRAMW6432': 'C:\\Program Files', 'USERNAME': 'sysadmin.uk', 'LOGONSERVER': '\\\\UK-DC1', 'PROMPT': '$P$G', 'COMSPEC': 'C:\\Windows\\system32\\cmd.exe', 'PROGRAMDATA': 'C:\\ProgramData', 'USERDNSDOMAIN': 'GLOBAL.LOCAL', 'PATHEXT': '.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC', 'FP_NO_HOST_CHECK': 'NO', 'WINDIR': 'C:\\Windows', 'OPENSSL_CONF': 'C:\\OpenSSL-Win64\\bin\\openssl.cfg', 'HOMEDRIVE': 'C:', 'APPDATA': 'C:\\Users\\sysadmin.uk\\AppData\\Roaming', 'SYSTEMDRIVE': 'C:', 'NUMBER_OF_PROCESSORS': '1', 'PROCESSOR_LEVEL': '6', 'COMMONPROGRAMW6432': 'C:\\Program Files\\Common Files', 'OS': 'Windows_NT', 'PUBLIC': 'C:\\Users\\Public', 'USERPROFILE': 'C:\\Users\\sysadmin.uk'}, 'stdout': -1, 'stdin': None, 'stderr': -2, 'cwd': 'C:\\Users\\sysadmin.uk'}, reason: [Error 5] Access is denied

Note far right end of line above reads:

reason: [Error 5] Access is denied
```
@twangboy @jfindlay can you please help me with this?

**Comment by jfindlay** [comment]
@TheBigBear, I'm not sure what the real issue is here.  It seems that there are some issues with the winrepo templating that need to be worked out.

**Comment by TheBigBear** [comment]
@twangboy I think this might be a jinja whietspace control issue as discussed here:
http://jinja.pocoo.org/docs/dev/templates/#whitespace-control
```
You can also strip whitespace in templates by hand. If you add a minus sign (-) to the start or end of a block (e.g. a For tag), a comment, or a variable expression, the whitespaces before or after that block will be removed:

{% for item in seq -%}
    {{ item }}
{%- endfor %}

This will yield all elements without whitespace between them. If seq was a list of numbers from 1 to 9, the output would be 123456789.
```
But I can't find the right place to insert this into salt and what I really don't get, is, why it works without adding a trailing space on 32-bit machines and only adds that trailing space on a 64-bit machine.

**Comment by twangboy** [comment]
This should be fixed with #26625. There was a portion of the code that was trying to remove the (x86). Now that we're using jinja, it's unnecessary. That should be merged today.
